src/kaleidoscope/layers.cpp

```diff
@@ -84,45 +84,43 @@ void Layer_::handleKeymapKeyswitchEvent(Key keymapEntry, uint8_t keyState) {
   }
 }
 
-Key Layer_::eventHandler(Key mappedKey, byte row, byte col, uint8_t keyState) {
+Key Layer_::eventHandler(Key mappedKey, KeyAddr key_addr, uint8_t keyState) {
   if (mappedKey.flags != (SYNTHETIC | SWITCH_TO_KEYMAP))
     return mappedKey;
 
   handleKeymapKeyswitchEvent(mappedKey, keyState);
   return Key_NoKey;
 }
 
-Key Layer_::getKeyFromPROGMEM(uint8_t layer, byte row, byte col) {
+Key Layer_::getKeyFromPROGMEM(uint8_t layer, KeyAddr key_addr) {
   Key key;
 
-  key.raw = pgm_read_word(&(keymaps[layer][row][col]));
+  key.raw = pgm_read_word(&(keymaps[layer][key_addr.row()][key_addr.col()]));
 
   return key;
 }
 
-void Layer_::updateLiveCompositeKeymap(byte row, byte col) {
-  int8_t layer = active_layers_[row][col];
-  live_composite_keymap_[row][col] = (*getKey)(layer, row, col);
+void Layer_::updateLiveCompositeKeymap(KeyAddr key_addr) {
+  int8_t layer = active_layers_[key_addr.toInt()];
+  live_composite_keymap_[key_addr.toInt()] = (*getKey)(layer, key_addr);
 }
 
 void Layer_::updateActiveLayers(void) {
-  memset(active_layers_, 0, ROWS * COLS);
-  for (byte row = 0; row < ROWS; row++) {
-    for (byte col = 0; col < COLS; col++) {
-      int8_t layer = top_active_layer_;
-
-      while (layer > 0) {
-        if (Layer.isActive(layer)) {
-          Key mappedKey = (*getKey)(layer, row, col);
-
-          if (mappedKey != Key_Transparent) {
-            active_layers_[row][col] = layer;
-            break;
-          }
+  memset(active_layers_, 0, KeyboardHardware.numKeys());
+  for (auto key_addr : allKeyAddr()) {
+    int8_t layer = top_active_layer_;
+
+    while (layer > 0) {
+      if (Layer.isActive(layer)) {
+        Key mappedKey = (*getKey)(layer, key_addr);
+
+        if (mappedKey != Key_Transparent) {
+          active_layers_[key_addr.toInt()] = layer;
+          break;
         }
-        layer--;
       }
     }
+    layer--;
```

Comment:
This decrement needs to be inside the `while` loop, or it never terminates, and the keyboard hangs whenever a layer key is pressed.

Reply:
Thanks.